README.md
RiskBattler
===========

Throw ncurses dice and speed up your Risk game!

---

Files
-----
 - dice.py: Takes a number in [1,6] and creates an ascii art die. Also has the ability to join two ascii art dice in the same line.
 - risk.py: Main ncurses program.

Using this program
------------------
 1. Set up your Risk game like usual.
 2. Put away the dice.
 3. Start this program (python 2.7 or 3.x).
 4. When a player attacks, enter the troops they are using as offensive count.
 5. Start the battle with "space" to automatically finish the battle OR any other key to battle by one roll at time.

Known Issues
------------
 - The program requires 21 console lines to run properly.


CHANGELOG
BUILD: 0.1 alpha
NOTES: Initial setup, basic IO

BUILD: 1.0 alpha
NOTES: Initial release, ncurses

    BUILD: 1.0.1 alpha
    FIXED: Made sure str(battle) is displayed before user
             makes choice to auto play or not

    BUILD: 1.0.2 alpha
    FIXED: Made input more clear as to when you can quit
             the program and when you can break a battle

    BUILD: 1.0.3 alpha
    FIXED: Moved result line below Battle string

    BUILD: 1.0.4 alpha
    FIXED: Minimize writes to screen (optimization)

BUILD: 2.0 alpha
FIXED: First release of Three Offense Vs Two Defense
         dice system
NOTES: Need to check rules to avoid Offense troops
         falling to zero
       Need to replace assert statements

    BUILD: 2.1.1 alpha
    FIXED: Make sure all user input is checked for exit

    BUILD: 2.1.2 alpha
    FIXED: Use the correct number of dice for 3, 2, and 1
             attacking Armies

    BUILD: 2.1.3 alpha
    FIXED: Finally removed assert statements from risk.py

BUILD: 3.0 Beta
FIXED: Changed to a horazontal layout. Highest die on the
         left
       Corrected an error where a valid defense die was
         not displayed because it was tied to the offensive
         die count
       Lowered the required line count to twenty one lines


dice.py
#!/usr/bin/env python3
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301, USA.

"""Never lose the dice again!

The default is to print dice with `chr(183)` as the dots,
this can be changed by giving the first argument as a single character.
"""

def join(*faces):
    faces = [face.split("\n") for face in faces]

    lines = ['' for _ in faces[0]]
    for face in faces:
        for line in range(len(face)):
            lines[line] += face[line]

    return "\n".join(lines)

def str(number, dice_dots=chr(183)):
    """Returns the multi-line str relivant to number"""
    line = " | {0:^5} | "
    top = bottom = "  -------  "

    s = [top] + [line.format(x) for x in side(number)] + [bottom]

    return ("\n".join(s)).replace("x", dice_dots)

def side(number):
    """Returns the dots (x) relivant to number"""
    one = ["     ", "  x  ", "     "]
    two = ["x    ", "     ", "    x"]
    three = ["x    ", "  x  ", "    x"]
    four = ["x   x", "     ", "x   x"]
    five = ["x   x", "  x  ", "x   x"]
    six = ["x   x", "x   x", "x   x"]
    return [one, two, three, four, five, six][(number - 1)]

if __name__ == "__main__":
    raise RuntimeError("dice.py is a module")


risk.py
#!/usr/bin/env python3
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301, USA.

"""Speeds up the battle part of risk

By default it will roll until a winner is found
You have the option to go by one attack at a time so
you can stop the battle before you lose your advantage
"""

__all__ = ['main', 'get_int', 'Army', 'Battle']
__author__ = "Cody A. Taylor ( codemister99@yahoo.com )"
__version__ = "2.1.3 Alpha"

import curses
import dice
from random import randint

class Army:
    """An army of troops

    Keeps track of the number of troops you have
    As well as the roll for each individual attack
    """
    def __init__(self, troops, reserve=0):
        """Initializes the army with the given number of troops

        Sets roll to Zero to indicate no attack has occured
        """
        self.__start = troops
        self.__troops = troops
        self.__reserve = reserve

    def __repr__(self):
        return "Army(troops={!r}, reserve={!r})".format(self.__troops, self.__reserve)

    def __str__(self):
        return "{} troops; {} lost".format(self.__troops, self.lost())

    def lose(self, troops=1):
        """Lose the given number of troops

        By default One troop is lost
        Returns self for conveince
        """
        self.__troops -= troops
        return self

    def roll(self, count=1):
        """Set the roll to a new random number [1, 6]

        count: The count of dice to roll
        reserve: The the of troops to exclude

        Returns a list of sorted random numbers
        """
        self.__roll = []

        # This sets the correct count for dwindled armies.
        if count >= self.__troops:
            count = self.__troops - self.__reserve

        if count < 1:
            raise ValueError("The count of dice to roll is less than 1. " \
                "You did some voodoo! Seriously though, this should never " \
                "happen and you should report a bug.")

        return sorted([randint(1, 6) for roll in range(count)], reverse=True)

    def lost(self):
        """The number of troops lost since start of battle"""
        return self.__start - self.__troops

    def can_attack(self):
        """True when this army can still attack"""
        return self.__troops > 1

    def can_defend(self):
        """True when this army can still defend"""
        return self.__troops > 0

class Battle:
    """A Battle between two Armies"""
    def __init__(self, offense, defense):
        self.__offense = offense
        self.__defense = defense

    def __repr__(self):
        return "Battle(offense={!r}, defense={!r})".format(self.__offense, self.__defense)

    def __str__(self):
        return "Offense: {0}\nDefense: {1}".format(self.__offense, self.__defense)

    def action(self):
        """True if the Offense can still attack and the defense can still defend"""
        return self.__offense.can_attack() and self.__defense.can_defend()

    def attack(self):
        """Roll for both armies, lose troops accordingly

        Returns (offense_roll, defense_roll, str_result)
        """
        o_roll = self.__offense.roll(3)
        d_roll = self.__defense.roll(2)
        result = ""

        if o_roll[0] > d_roll[0]:
            self.__defense.lose()
            result += "Won"
        else:
            self.__offense.lose()
            result += "Lost"

        if len(o_roll) > 1 and len(d_roll) > 1:
            if o_roll[1] > d_roll[1]:
                self.__defense.lose()
                result += " and Won"
            else:
                self.__offense.lose()
                result += " and Lost"

        return o_roll, d_roll, result

    @property
    def offense(self):
        return self.__offense

    @property
    def defense(self):
        return self.__defense

def get_input(method):
    """Checks the result of 'method' to see if we should exit"""
    ret = method()
    if ret in (b'q', b'Q', ord('q'), ord('Q')):
        raise SystemExit
    return ret

def get_int(scr, x, msg):
    """Basic ncurses get_int

    clears away any text it writes before returning
    allow the use of the entire line (x)
    so that the user has plenty of room for input
    """
    curses.echo()
    curses.curs_set(1)
    scr.addstr(x, 0, "{}: ".format(msg))
    scr.refresh()

    i = ""
    while type(i) != int:
        i = get_input(scr.getstr)
        try:
            i = int(i)
        except ValueError:
            scr.addstr(x, len(msg) + 2, " " * len(i))
            scr.move(x, len(msg) + 2)

    scr.addstr(x, 0, " " * (len(msg) + 2 + len(str(i))))
    curses.noecho()
    curses.curs_set(0)
    return i

def main(stdscr):
    curses.curs_set(0)

    stdscr.addstr(0, 20, "Welcome to the Risk Battler!")

    while True:
        offense = get_int(stdscr, 2, "Number of offensive troops")
        defense = get_int(stdscr, 2, "Number of defensive troops")

        stdscr.addstr(2, 0, "Hit ' ' to automatically finish battles, 'q' to quit, any other for manual play")

        battle = Battle(Army(offense, 1), Army(defense))

        stdscr.refresh()
        ch = get_input(stdscr.getch)

        auto = True if ch == ord(' ') else False

        stdscr.addstr(2, 0, " " * 80)
        if auto == False:
            stdscr.addstr(2, 0, "Hit ' ' to continue battle, 'q' to quit, any other to stop")

        while battle.action():
            stdscr.refresh()

            if auto == False:
                ch = get_input(stdscr.getch)
                if ch != ord(' '):
                    break
            else:
                curses.napms(600)

            o_roll, d_roll, result = battle.attack()
            o_roll = [dice.str(d) for d in o_roll]
            d_roll = [dice.str(d) for d in d_roll]

            stdscr.addstr(4, 2, "Offense: {:>20}".format(str(battle.offense)))
            stdscr.addstr(6, 0, "\n".join([" " * 33] * 5))
            stdscr.addstr(6, 0, dice.join(*o_roll))

            stdscr.addstr(12, 2, "Defense: {:>20}".format(str(battle.defense)))
            stdscr.addstr(14, 0, "\n".join([" " * 22] * 5))
            stdscr.addstr(14, 0, dice.join(*d_roll))

            stdscr.addstr(20, 0, "Offense roll {:<13}".format(result))

        if auto == False:
            stdscr.addstr(2, 0, " " * 80)

        stdscr.refresh() # must refresh one last time to show final result

    return 0

if __name__ == "__main__":
    curses.wrapper(main)


riskc.py
#!/usr/bin/env python3
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
# MA 02110-1301, USA.

"""Speeds up the battle part of risk

By default it will roll until a winner is found
You have the option to go by one attack at a time so
you can stop the battle before you lose your advantage
"""

__all__ = ['main', 'get_int', 'Army', 'Battle']
__author__ = "Cody A. Taylor ( codemister99@yahoo.com )"
__version__ = "2.1 Alpha"

import curses
import dice
from random import randint

class Army:
    """An army of troops

    Keeps track of the number of troops you have
    As well as the roll for each individual attack
    """
    def __init__(self, troops):
        """Initializes the army with the given number of troops

        Sets roll to Zero to indicate no attack has occured
        """
        self.__troops = troops
        self.__roll = []

    def __repr__(self):
        return "{0!r} troops".format(self.__troops)

    def __str__(self):
        return "{} troops".format(self.__troops)

    def __eq__(self, other):
        """Compares number of troops"""
        return self.__troops == other.__troops

    def __ne__(self, other):
        """Compares number of troops"""
        return self.__troops != other.__troops

    def __lt__(self, other):
        """Compares last roll"""
        return self.__roll < other.__roll

    def __le__(self, other):
        """Compares last roll"""
        return self.__roll <= other.__roll

    def __gt__(self, other):
        """Compares last roll"""
        return self.__roll > other.__roll

    def __ge__(self, other):
        """Compares last roll"""
        return self.__roll >= other.__roll

    def lose(self, troops=1):
        """Lose the given number of troops

        By default One troop is lost
        Returns self for conveince
        """
        self.__troops -= troops
        return self

    def roll(self, number=1):
        """Set the roll to a new random number [1, 6]

        Returns a list of sorted random numbers
        """
        self.__roll = []

        if number > self.__troops:
            number = self.__troops

        for roll in range(number):
            self.__roll.append(randint(1, 6))

        self.__roll.sort()
        self.__roll.reverse()

        return self.__roll

    def can_attack(self):
        """True when this army can still attack"""
        return self.__troops > 1

    def can_defend(self):
        """True when this army can still defend"""
        return self.__troops > 0

class Battle:
    """A Battle between two Armies"""
    def __init__(self, offense, defense):
        self.__offense = offense
        self.__defense = defense

    def __repr__(self):
        return "Offense: {0!r}    \nDefense: {1!r}    ".format(self.__offense, self.__defense)

    def __str__(self):
        return "Offense: {0}    \nDefense: {1}    ".format(self.__offense, self.__defense)

    def action(self):
        """True if the Offense can still attack and the defense can still defend"""
        return self.__offense.can_attack() and self.__defense.can_defend()

    def attack(self):
        """Roll for both armies, lose troops accordingly

        Returns (offense_roll, defense_roll, str_result)
        """
        o_roll = self.__offense.roll(3)
        d_roll = self.__defense.roll(2)
        result = ""

        assert "No die for offense", len(o_roll) == 0
        assert "No die for defense", len(d_roll) == 0

        if o_roll[0] > d_roll[0]:
            self.__defense.lose()
            result += "Won "
        else:
            self.__offense.lose()
            result += "Lost "

        if len(o_roll) > 1 and len(d_roll) > 1:
            if o_roll[1] > d_roll[1]:
                self.__defense.lose()
                result += "and Won"
            else:
                self.__offense.lose()
                result += "and Lost"

        return o_roll, d_roll, result

def get_int(scr, x, msg):
    """Basic ncurses get_int

    clears away any text it writes before returning
    allow the use of the entire line (x)
    so that the user has plenty of room for input
    """
    curses.echo()
    curses.curs_set(1)
    scr.addstr(x, 0, "{}: ".format(msg))
    scr.refresh()

    i = ""
    while type(i) != int:
        i = scr.getstr()
        try:
            i = int(i)
        except ValueError:
            scr.addstr(x, len(msg) + 2, " " * len(i))
            scr.move(x, len(msg) + 2)

    scr.addstr(x, 0, " " * ( len(msg) + 2 + len(str(i)) ) )
    curses.noecho()
    curses.curs_set(0)
    return i

def main(stdscr):
    curses.curs_set(0)

    stdscr.addstr(0, 20, "Welcome to the Risk Battler!")
    stdscr.addstr(2, 0, "  Offense")
    stdscr.addstr(3, 0, dice.join(dice.str(1), dice.str(1), dice.str(1)))
    stdscr.addstr(8, 0, "  Defense")
    stdscr.addstr(9, 0, dice.join(dice.str(1), dice.str(1)))

    ch = ord('1')
    while ch != ord('q'):
        offense = get_int(stdscr, 1, "Number of offensive troops")
        defense = get_int(stdscr, 1, "Number of defensive troops")

        stdscr.addstr(1, 0, "Hit ' ' to automatically finish battles, 'q' to quit, any other for manual play")

        battle = Battle(Army(offense), Army(defense))
        stdscr.addstr(14, 0, str(battle))

        stdscr.refresh()
        ch = stdscr.getch()

        auto = True if ch == ord(' ') else False

        if auto == False:
            stdscr.addstr(1, 0, "Hit ' ' to continue battle, 'q' to quit, any other to stop                     ")
        else:
            stdscr.addstr(1, 0, "                                                                               ")

        while battle.action() and ch != ord('q'):
            stdscr.refresh()

            if auto == False:
                ch = stdscr.getch()
                if ch != ord(' '):
                    break
            else:
                curses.napms(600)

            o_roll, d_roll, result = battle.attack()

            stdscr.addstr(3, 0, "{:31}\n".format(" ") * 5)
            stdscr.addstr(3, 0, dice.join(*[dice.str(r) for r in o_roll]))
            stdscr.addstr(9, 0, "{:20}\n".format(" ") * 5)
            stdscr.addstr(9, 0, dice.join(*[dice.str(r) for r in d_roll]))
            stdscr.addstr(14, 0, str(battle))
            stdscr.addstr(16, 0, "Offense roll {}          ".format(result))

        if auto == False:
            stdscr.addstr(1, 0, "                                                          ")

        stdscr.refresh() # must refresh one last time to show final result

    return 0

if __name__ == "__main__":
    curses.wrapper(main)
